Reworked error output to send IETF RFC-7807 compatible problem responses

Alexamakans opened the pull request:
Closes #13
Depends on https://github.com/iver-wharf/wharf-core/pull/17

Alexamakans committed:
Changed to use wharf-core/pkg/ginutil for error responses

Alexamakans committed:
Readded accidentally removed fmt.Println

Alexamakans committed:
Readded accidentally removed fmt.Println

Also fixed indentation

Alexamakans committed:
Merge branch 'feature/use-wharf-core-ginutil' of github.com:iver-wharf/wharf-provider-azuredevops into feature/use-wharf-core-ginutil

Alexamakans committed:
Fix indentation

Alexamakans committed:
Updated CHANGELOG.md

Alexamakans committed:
Changed Failure response type

applejag commented on azuredevops.go:
```diff
@@ -93,16 +93,21 @@ func runAzureDevOpsHandler(c *gin.Context) {
-		c.Error(err)
-		c.JSON(http.StatusBadRequest, err.Error())
+		ginutil.WriteInvalidBindError(c, err,
+			"One or more parameters failed to parse when reading the request body for branch object to update.")
```
This seems a little copy-pasted

```suggestion
			"One or more parameters failed to parse when reading the request body for import details.")
```

applejag commented on azuredevops.go:
```diff
@@ -93,16 +93,21 @@ func runAzureDevOpsHandler(c *gin.Context) {
+			Title: "Empty group error.",
+			Status: http.StatusBadRequest,
+			Detail: "Unable to get due to empty group.",
+		})
```
Suggest reusing something like the [`/prob/api/invalid-param`](https://iver-wharf.github.io/#/prob/api/invalid-param) here instead, perhaps by using the [`WriteInvalidParamError`](https://github.com/iver-wharf/wharf-core/blob/c2f9a20c6a640196d4518573925f154cba982b97/pkg/ginutil/writeproblem.go#L128-L140)

applejag commented on azuredevops.go:
```diff
@@ -111,22 +116,37 @@ func runAzureDevOpsHandler(c *gin.Context) {
+				Title: "Error getting token.",
+				Status: http.StatusBadRequest,
+				Detail: "Unable to get token.",
+			})
```
We later want to use the problem responses taken from the wharf-api-client-go (as used in `client.GetProviderById` above), which will have more descriptive problems in certain scenarios.

In the meantime, suggest generalizing this to something like `/prob/provider/unexpected-api-read-error`, as we could use that as fallback if there's some non-problem error from the wharf-api-client-go.

applejag commented on azuredevops.go:
```diff
@@ -111,22 +116,37 @@ func runAzureDevOpsHandler(c *gin.Context) {
+			Type: "prob/provider/azuredevops/empty-user-error",
+			Title: "Empty user error.",
+			Status: http.StatusBadRequest,
+			Detail: "Unable to get due to empty user.",
```
Suggest reusing something like the [`/prob/api/invalid-param`](https://iver-wharf.github.io/#/prob/api/invalid-param) here instead, perhaps by using the [`WriteInvalidParamError`](https://github.com/iver-wharf/wharf-core/blob/c2f9a20c6a640196d4518573925f154cba982b97/pkg/ginutil/writeproblem.go#L128-L140)

Perhaps with a `Detail` of `"Unable to import when both user and token are omitted."`

applejag commented on azuredevops.go:
```diff
@@ -158,13 +188,23 @@ func runAzureDevOpsHandler(c *gin.Context) {
+			Title: "Error getting body from request.",
+			Status: http.StatusBadRequest,
+			Detail: "Unable to get body from request.",
+		})
```
Suggest reusing [`WriteBodyReadError`](https://github.com/iver-wharf/wharf-core/blob/c2f9a20c6a640196d4518573925f154cba982b97/pkg/ginutil/writeproblem.go#L58-L69)

applejag commented on azuredevops.go:
```diff
@@ -158,13 +188,23 @@ func runAzureDevOpsHandler(c *gin.Context) {
+			Title: "Error building url.",
+			Status: http.StatusBadRequest,
+			Detail: "Unable to build url.",
+		})
```
Suggest reusing [`WriteInvalidParamError`](https://github.com/iver-wharf/wharf-core/blob/c2f9a20c6a640196d4518573925f154cba982b97/pkg/ginutil/writeproblem.go#L128-L140) instead

applejag commented on azuredevops.go:
```diff
@@ -182,22 +222,37 @@ func runAzureDevOpsHandler(c *gin.Context) {
+			Title: "Error unmarshalling project.",
+			Status: http.StatusBadRequest,
+			Detail: "Unable to unmarshal projects.",
+		})
```
This should definitely be a "502 Bad Gateway". Perhaps generalizing it to `/prob/provider/unexpected-response-format`, `.../unexpected-remote-response-format` ? I.e. use the same type throughout the different providers, where the problem description could say something like "Could be caused by invalid JSON data structure as a result of incompatible version of the Azure DevOps / GitLab / *et.al* API"

applejag commented on azuredevops.go:
```diff
@@ -182,22 +222,37 @@ func runAzureDevOpsHandler(c *gin.Context) {
 		if err != nil {
 			fmt.Println("Unable to get build definition: ", err)
-			c.JSON(http.StatusBadRequest, fmt.Sprintf("Error: %+v", err))
+			ginutil.WriteProblemError(c, err, problem.Response{
+				Type: "prob/provider/azuredevops/getting-build-definition-error",
+				Title: "Error getting build definition.",
+				Status: http.StatusBadRequest,
+				Detail: "Unable to get build definition.",
+			})
```
Suggest generalizing to something provider-agnostic like `/prob/provider/fetch-build-definition`

Adding `-error` to the type is unnecessary. It's implicit that it's an error or similar as it's a problem definition.

applejag commented on azuredevops.go:
```diff
@@ -182,22 +222,37 @@ func runAzureDevOpsHandler(c *gin.Context) {
 		if err != nil {
 			fmt.Println("Unable to construct git url ", err)
-			c.JSON(http.StatusBadRequest, fmt.Sprintf("Error: %+v", err))
+			ginutil.WriteProblemError(c, err, problem.Response{
+				Type: "prob/provider/azuredevops/constructing-git-url-error",
+				Title: "Error constructing git url.",
+				Status: http.StatusBadRequest,
+				Detail: "Unable to construct git url.",
+			})
```
Suggest generalizing to something like `/prob/provider/composing-provider-data`

applejag commented on azuredevops.go:
```diff
@@ -276,42 +341,47 @@ func runAzureDevOpsHandler(c *gin.Context) {
-		c.JSON(
-			http.StatusBadRequest,
-			fmt.Sprintf("Expected git.pullrequest.created trigger, got: %s instead.", t.EventType))
+		err := fmt.Errorf("expected git.pullrequest.created trigger, got: %s", t.EventType)
+		ginutil.WriteProblemError(c, err, problem.Response{
+			Type: "prob/provider/azuredevops/invalid-event-type-error",
+			Title: "Invalid event type.",
+			Status: http.StatusBadRequest,
+			Detail: "Invalid event type.",
+		})
```
The event type is probably valid, although it's unsupported. Suggest changing to something like `/prob/provider/azuredevops/unsupported-event-type`

Suggest to also extract the event type as a local const and then include the value in the detail as well.

```go
const eventTypePullRequest = "git.pullrequest.created"

// ...

Detail: fmt.Sprintf("Received event type %q, while only %q is supported.", t.EventType, eventTypePullRequest),
```

applejag commented on azuredevops.go:
```diff
@@ -111,22 +116,37 @@ func runAzureDevOpsHandler(c *gin.Context) {
-				fmt.Println("Unable to put token: ", err)
-				c.JSON(http.StatusBadRequest, fmt.Sprintf("Error: %+v", err))
+				fmt.Println("Unable to post token: ", err)
+				ginutil.WriteProblemError(c, err, problem.Response{
+					Type: "prob/provider/azuredevops/posting-token-error",
+					Title: "Error posting token.",
+					Status: http.StatusBadRequest,
+					Detail: "Unable to post token.",
+				})
```
Same as https://github.com/iver-wharf/wharf-provider-azuredevops/pull/14#discussion_r651834695, but `/prob/provider/unexpected-api-write-error`

applejag commented on azuredevops.go:
```diff
@@ -138,7 +158,12 @@ func runAzureDevOpsHandler(c *gin.Context) {
 		if err != nil || provider.ProviderID == 0 {
 			fmt.Printf("Unable to get provider. %+v", err)
-			c.JSON(http.StatusBadRequest, fmt.Sprintf("Unable to get provider. %+v", err))
+			ginutil.WriteProblemError(c, err, problem.Response{
+				Type: "prob/provider/azuredevops/getting-provider-error",
+				Title: "Error getting provider.",
+				Status: http.StatusBadRequest,
+				Detail: "Unable to get provider.",
+			})
```
Same as https://github.com/iver-wharf/wharf-provider-azuredevops/pull/14#discussion_r651834695

applejag commented on azuredevops.go:
```diff
@@ -147,8 +172,13 @@ func runAzureDevOpsHandler(c *gin.Context) {
-				fmt.Println("Unable to put provider: ", err)
-				c.JSON(http.StatusBadRequest, fmt.Sprintf("Error: %+v", err))
+				fmt.Println("Unable to post provider: ", err)
+				ginutil.WriteProblemError(c, err, problem.Response{
+					Type: "prob/provider/azuredevops/posting-provider-error",
+					Title: "Error posting provider.",
+					Status: http.StatusBadRequest,
+					Detail: "Unable to post provider.",
+				})
```
Same as https://github.com/iver-wharf/wharf-provider-azuredevops/pull/14#discussion_r651834695, but `/prob/provider/unexpected-api-write-error`

applejag commented on azuredevops.go:
```diff
@@ -213,7 +268,12 @@ func runAzureDevOpsHandler(c *gin.Context) {
 		if err != nil {
 			fmt.Println("Unable to put project: ", err)
-			c.JSON(http.StatusBadRequest, fmt.Sprintf("Error: %+v", err))
+			ginutil.WriteProblemError(c, err, problem.Response{
+				Type: "prob/provider/azuredevops/putting-project-error",
+				Title: "Error putting project.",
+				Status: http.StatusBadRequest,
+				Detail: "Unable to put project.",
+			})
```
Same as https://github.com/iver-wharf/wharf-provider-azuredevops/pull/14#discussion_r651834695, but `/prob/provider/unexpected-api-write-error`

applejag commented on azuredevops.go:
```diff
@@ -259,7 +319,12 @@ func runAzureDevOpsHandler(c *gin.Context) {
 			if err != nil {
 				fmt.Println("Unable to put branch: ", err)
-				c.JSON(http.StatusBadRequest, fmt.Sprintf("Error: %+v", err))
+				ginutil.WriteProblemError(c, err, problem.Response{
+					Type: "prob/provider/azuredevops/putting-branch-error",
+					Title: "Error putting branch.",
+					Status: http.StatusBadRequest,
+					Detail: "Unable to put branch.",
+				})
```
Same as https://github.com/iver-wharf/wharf-provider-azuredevops/pull/14#discussion_r651834695, but `/prob/provider/unexpected-api-write-error`

applejag commented on azuredevops.go:
```diff
@@ -320,9 +390,15 @@ func prCreatedTriggerHandler(c *gin.Context) {
+	if err != nil {
+		fmt.Println("Unable to send trigger to wharf-client: ", err)
+		err = fmt.Errorf("unable to send trigger to wharf-client: %v", err)
+		ginutil.WriteProblemError(c, err, problem.Response{
+			Type: "prob/provider/azuredevops/send-trigger-error",
+			Title: "Error sending trigger.",
+			Status: http.StatusBadRequest,
+			Detail: "Unable to send trigger to wharf-client.",
+		})
```
Same as https://github.com/iver-wharf/wharf-provider-azuredevops/pull/14#discussion_r651834695, but `/prob/build/unexpected-trigger-error`

fredx30 reviewed (dismissed):
@jilleJr Makes some good points. After those comments are resolved this has my 👍 .

This also no longer seems like a simple reference swap that the title hints at. I think it would be fair to call it a full rework of error output for the entire repo.

Alexamakans committed:
Refactored and changed problem responses to use more generic types

- Sending error responses has mostly been moved from runAzureDevOpsHandler to the functions that it calls instead.
- The name of functions that write an error to the gin.Context has the suffix -WritesError.
- Added some helpers in the form of extensions of existing code. (Not sure about this decision)
  - ginutilext, wharfapiext, requests

It still feels really messy. I hope I didn't make it worse :&

Alexamakans committed:
Fixed lint errors (Comments)

Alexamakans committed:
Changed which error was sent on requests.GetAndParseJSON errors

applejag commented on azuredevops.go:
```diff
@@ -1,24 +1,28 @@
 const apiProjects string = "_apis/projects"
+const apiProviderName string = "azuredevops"
+const apiVersion string = "5.0"
+const eventTypePullRequest string = "git.pullrequest.created"
```
Some of these consts are only used in specific endpoints. Suggest moving this specific const, `eventTypePullRequest`, into the `prCreatedTriggerHandler` function instead.

The `apiVersion` is highly function related as well, as different endpoints may use different AzureDevOps versions

applejag commented on azuredevops.go:
```diff
@@ -276,202 +164,386 @@ func runAzureDevOpsHandler(c *gin.Context) {
 
-func getBodyFromRequest(user string, token string, url string) ([]byte, error) {
-	req, err := http.NewRequest(http.MethodGet, url, nil)
+// tryPutProjectWritesError Writes an error to the gin Context when an error occurs!
```
Very nice of you to add documentation :)

There's no real need here though, as we can much more easily just look at the implementation as it's in the same code base. I.e. the guidelines for when to add comments is mostly "only for exported types". Internal comments usually quickly get outdated. The fewer places to have to maintain the better.

applejag commented on azuredevops.go:
```diff
@@ -59,6 +68,11 @@ type azureDevOpsRepository struct {
+type azureDevOpsRepositorySlice struct {
+	Count int `json:"count"`
+	Value []azureDevOpsRepository `json:"value"`
+}
```
This type is not a slice itself, but contains a value which is a slice.

Suggest renaming it to something like:

- azureDevOpsRepositoryResponse
- azureDevOpsRepositories

Same argument for azureDevOpsProjectSlice above

applejag commented on helpers/ginutilext/extension.go:
```diff
@@ -0,0 +1,100 @@
+package ginutilext
```
Suggest moving this package and this file name

`/helpers/ginutilext/extension.go` &rarr; `/pkg/ginutilext/ginutilext.go`
or 
`/helpers/ginutilext/extension.go` &rarr; `/pkg/ginutilext/writeproblem.go`

"helpers" is unconventional Go naming. A summary of folder names based on conventions, which we've used as a source of truth while learning Go :), is this repo: https://github.com/golang-standards/project-layout

applejag commented on helpers/ginutilext/extension.go:
```diff
@@ -0,0 +1,100 @@
+	})
+}
+
+// WriteAuthenticationError uses WriteProblemError to write a 401 "Unauthorized"
+// response with the type "prob/provider/authentication-error".
+//
+// Meant to be used for failed authentication.
+func WriteAuthenticationError(c *gin.Context, err error, detail string) {
+	ginutil.WriteProblemError(c, err, problem.Response{
+		Type: "prob/provider/authentication-error",
+		Title: "Error authenticating.",
+		Status: http.StatusUnauthorized,
+		Detail: detail,
+	})
+}
```
This functions seems unused. Suggest removing it.

We will need it in the future, but I'd name this premature.

Alexamakans replied:
Related to: https://github.com/iver-wharf/wharf-provider-github/pull/16#discussion_r655261660

This one is currently used in my PR for GitLab.
Thinking about it now maintaining the same file over 3 repos feels foolish.

What do you think is best?
1. Have it like it is, but only have the relevant functions in each repo. 
2. Submit a PR to wharf-core, and having the provider repo PRs depend on it.

applejag replied:
Vote for 2.

applejag commented on helpers/requests/requests.go:
```diff
@@ -0,0 +1,87 @@
+package requests
```
Same here:

`/helpers/requests/requests.go` &rarr; `/pkg/requests/requests.go`

applejag commented on helpers/requests/requests.go:
```diff
@@ -0,0 +1,87 @@
+	"net/url"
+)
+
+// ConstructGetURL Constructs a URL to use in a GET request.
+// Queries are passed as a map of string arrays
```
The function name is a word in the sentence. No need to capitalize the word "Constructs" here.

Also, this creates a kind of double-wording. Suggest using "creates" instead.

The second sentence is sort of implied from the function signature IMO.

```suggestion
// ConstructGetURL creates a GET request from a base URL, a map of query
// parameters, and a formatted path.
```

applejag commented on helpers/requests/requests.go:
```diff
@@ -0,0 +1,87 @@
+	rawURL string, queries map[string][]string, format string, values ...interface{}) (*url.URL, error) {
+
+	urlPath, err := url.Parse(rawURL)
+	if err != nil {
+		fmt.Printf("Unable to parse url %q", rawURL)
+		return nil, err
+	}
```
Excessive logging (IMO). Better to only return the error and let the user of this function decide when to log in this case.

```suggestion
	if err != nil {
		return nil, err
	}
```

applejag commented on helpers/requests/requests.go:
```diff
@@ -0,0 +1,87 @@
+	return urlPath, nil
+}
+
+// GetAndParseJSON Makes a request to urlPath with basic auth using user and token.
+//	Attempts to parse the JSON response, returning it in "result" if successful.
```
nit: Not returning anything. You're altering the true value

Also, I've found that referring to parameter names in the comments is not that sustainable as barely any refactoring tool will update the names there. Suggest to instead refer to their meaning/usage instead. Here we could even skip the `urlPath` as it's IMO obvious what it's used for.

Also2, basic auth is always user+password.

```suggestion
// GetAndParseJSON makes a HTTP request with basic auth.
// On success the response body will be unmarshalled as JSON.
```

applejag commented on helpers/requests/requests.go:
```diff
@@ -0,0 +1,87 @@
+	return string(body), nil
+}
+
+func getBodyFromRequest(user string, token string, url string) ([]byte, error) {
```
You're always calling this method with an `*url.URL`. Suggest letting the `url` parameter be a `*url.URL` and calling `.String()` inside this function instead

applejag commented on helpers/requests/requests.go:
```diff
@@ -0,0 +1,87 @@
+	return nil
+}
+
+// GetAsString Makes a request to urlPath with basic auth using user and token.
+// Returns the response as a string
```
I've found that referring to parameter names in the comments is not that sustainable as barely any refactoring tool will update the names there. Suggest to instead refer to their meaning/usage instead. Here we could even skip the `urlPath` as it's IMO obvious what it's used for.

```suggestion
// GetAsString invokes a HTTP request with basic auth.
// Returns the response as a string.
```

applejag commented on helpers/wharfapiext/extension.go:
```diff
@@ -0,0 +1,181 @@
+}
+
+// CreateOrUpdate Convenience function to POST or PUT data to a Wharf endpoint.
+func (c ExtClient) CreateOrUpdate(data interface{}, method, dataName, endpoint string) (interface{}, error) {
+	var newData = reflect.New(reflect.TypeOf(data)).Elem().Interface()
```
Strongly advice against using reflection. Here you're trying to make a generic function in a language that does not support generics. Reflection is forever slow, and the casting required is unsafe and error-prone.

Suggest looking into alternative solutions. Such as just returning the response body bytes as you're doing in your `requests` package instead.

applejag commented on helpers/wharfapiext/extension.go:
```diff
@@ -0,0 +1,181 @@
+	log "github.com/sirupsen/logrus"
+)
+
+// ExtClient Extension of the wharfapi.Client
+type ExtClient struct {
+	*wharfapi.Client
+}
```
We do control the `wharfapi` library, so I suggest moving your code over there, in as appropriate way as you can imagine without draggin the AzureDevOps domain into it.

Looks like you've copy-pasted a lot of code from the `wharfapi` library itself. Even introduced logrus from there.

Sorry, but I suggest rethinking this `wharfapiext` package. :(

Alexamakans replied:
I don't quite remember why I even created it, so that's a fair suggestion.
It doesn't seem to be anything other than some refactoring of existing functions.

I will remove it completely.

applejag reviewed (changes requested):
Did not review it all, so will have to revisit. But I got a few notes

Alexamakans committed:
Removed ExtClient

Alexamakans committed:
Renamed helpers -> pkg

Alexamakans committed:
Updated to use wharf-core v1.0.0

Alexamakans committed:
Changed to use wharf-core/pkg/ginutil instead of pkg/ginutilext

Alexamakans committed:
Updated go.mod and go.sum

Alexamakans committed:
Applied code review suggestions

Alexamakans committed:
Merge branch 'master' of github.com:iver-wharf/wharf-provider-azuredevops into feature/use-wharf-core-ginutil

Alexamakans committed:
Updated CHANGELOG.md

Alexamakans committed:
Updated CHANGELOG.md

fredx30 commented on pkg/requests/requests.go:
```diff
@@ -0,0 +1,78 @@
+
+	req.SetBasicAuth(user, token)
+
+	resp, err := http.DefaultClient.Do(req)
```
CodeQL warning here. This needs some more explanation/discussion/analysis in this PR.

fredx30 replied:
Warning here - https://github.com/iver-wharf/wharf-provider-azuredevops/security/code-scanning/8?query=ref%3Arefs%2Fpull%2F14%2Fmerge

Alexamakans replied:
Analysis here: https://github.com/iver-wharf/wharf-provider-azuredevops/pull/14#issuecomment-880575891

Alexamakans commented:
The CodeQL alert doesn't seem like a false positive. (https://github.com/iver-wharf/wharf-provider-azuredevops/security/code-scanning/8?query=ref%3Arefs%2Fpull%2F14%2Fmerge)

I couldn't find a place where i.URL was validated beyond being a string that passes url.Parse.
The full URL then seems to take the forms of
- `{A}/{B}/_apis/projects?api-version=5.0`
- `{A}/{B}/_apis/projects/{C}?api-version=5.0`
where
- `{A}` is a string that passes url.Parse.
- `{B}` is the group name passed in. Not sure if there are any validity checks for it.
- `{C}` is the project name passed in. Not sure if there are any validity checks for it.
used in a GET request meant for the remote provider.

I don't know enough about vulnerabilities in this area, so I'm not sure if this could cause any trouble or not.

applejag commented:
> The CodeQL alert doesn't seem like a false positive. (https://github.com/iver-wharf/wharf-provider-azuredevops/security/code-scanning/8?query=ref%3Arefs%2Fpull%2F14%2Fmerge)
> 
> I couldn't find a place where i.URL was validated beyond being a string that passes url.Parse.
> The full URL then seems to take the forms of
> 
> * `{A}/{B}/_apis/projects?api-version=5.0`
> * `{A}/{B}/_apis/projects/{C}?api-version=5.0`
>   where
> * `{A}` is a string that passes url.Parse.
> * `{B}` is the group name passed in. Not sure if there are any validity checks for it.
> * `{C}` is the project name passed in. Not sure if there are any validity checks for it.
>   used in a GET request meant for the remote provider.
> 
> I don't know enough about vulnerabilities in this area, so I'm not sure if this could cause any trouble or not.

As the first parameter comes from our code that would be safe. But I guess the provided code does not guarantee this in a reasonable way if say the `{A}` also is a user input, which it could be.

Don't know how in-depth the vulnerability scan is, but looking at what it flagged it seems that it's pretty strict. It seems like as long as you use even a functions argument in the request, then it will flag it as vulnerable. No matter how that function is later used.

What we could, and probably should do, is add a whitelist of base URLs in the config that the user feeds in lists of provider sources from.

Such as:

```yaml
# wharf-provider-azuredevops-config.yaml
providerPrefixWhitelist:
  - https://dev.azure.com/
```

Then that would resolve a lot of security issues with this implementation.

Alexamakans commented:
> > The CodeQL alert doesn't seem like a false positive. (https://github.com/iver-wharf/wharf-provider-azuredevops/security/code-scanning/8?query=ref%3Arefs%2Fpull%2F14%2Fmerge)
> > I couldn't find a place where i.URL was validated beyond being a string that passes url.Parse.
> > The full URL then seems to take the forms of
> > 
> > * `{A}/{B}/_apis/projects?api-version=5.0`
> > * `{A}/{B}/_apis/projects/{C}?api-version=5.0`
> >   where
> > * `{A}` is a string that passes url.Parse.
> > * `{B}` is the group name passed in. Not sure if there are any validity checks for it.
> > * `{C}` is the project name passed in. Not sure if there are any validity checks for it.
> >   used in a GET request meant for the remote provider.
> > 
> > I don't know enough about vulnerabilities in this area, so I'm not sure if this could cause any trouble or not.
> 
> As the first parameter comes from our code that would be safe. But I guess the provided code does not guarantee this in a reasonable way if say the `{A}` also is a user input, which it could be.
> 
> Don't know how in-depth the vulnerability scan is, but looking at what it flagged it seems that it's pretty strict. It seems like as long as you use even a functions argument in the request, then it will flag it as vulnerable. No matter how that function is later used.
> 
> What we could, and probably should do, is add a whitelist of base URLs in the config that the user feeds in lists of provider sources from.
> 
> Such as:
> 
> ```yaml
> # wharf-provider-azuredevops-config.yaml
> providerPrefixWhitelist:
>   - https://dev.azure.com/
> ```
> 
> Then that would resolve a lot of security issues with this implementation.

That feels like a good solution.
Am I correct in assuming that this should go in wharf-api?

applejag commented:
> That feels like a good solution.
> Am I correct in assuming that this should go in wharf-api?

With today's solution that can't live inside the wharf-api. When importing a project via wharf-web, you talk directly to the wharf-provider-azuredevops service. It would require an extra roundtrip then to be implemented inside the wharf-provider-azuredevops service (and others), while we also want to stick to DRY as much as possible.

Tricky topic that I think should be resolved in a different PR. Could you create an issue on this?

applejag commented on azuredevops.go:
```diff
@@ -1,24 +1,24 @@
 const apiProjects string = "_apis/projects"
+const apiProviderName string = "azuredevops"
 const itemsPath string = "items"
 const refsPath string = "refs"
```
nit: These are becomming so many that I suggest grouping them into a single `const` statement for extra readability as that aligns the values on `gofmt`, and perhaps dropping the `string` type as that's inferred

```suggestion
const (
	apiRepositories = "_apis/git/repositories"
	apiProjects     = "_apis/projects"
	apiProviderName = "azuredevops"
	itemsPath       = "items"
	refsPath        = "refs"
)
```

applejag commented on azuredevops.go:
```diff
@@ -48,6 +48,11 @@ type azureDevOpsProject struct {
+type azureDevOpsProjectResponse struct {
+	Count int `json:"count"`
+	Value []azureDevOpsProject `json:"value"`
+}
```
Seems to be invalid formatting. Suggest running `gofmt -w` or `goimports -w` on these files

applejag commented on azuredevops.go:
```diff
@@ -90,179 +101,52 @@ func runAzureDevOpsHandler(c *gin.Context) {
 	if i.Group == "" {
 		fmt.Println("Unable to get due to empty group.")
-		c.JSON(http.StatusBadRequest, "Unable to get due to empty group.")
+		ginutil.WriteInvalidParamError(c, err, "group",
+			"Unable to import due to empty group.")
```
`err` is `nil` here. As we don't have a `ginutil.WriteInvalidParam` without the Error suffix, you could create your own error here if you want. Such as:

```suggestion
		err := errors.New("missing required property: group")
		ginutil.WriteInvalidParamError(c, err, "group",
			"Unable to import due to empty group.")
```

applejag commented on azuredevops.go:
```diff
@@ -90,179 +101,52 @@ func runAzureDevOpsHandler(c *gin.Context) {
+	token, err := getOrPostTokenWritesError(c, client, i)
 	if err != nil {
-		fmt.Println("Unable to build url: ", err)
-		c.JSON(http.StatusBadRequest, fmt.Sprintf("Error: %+v", err))
+		fmt.Println("Unable to get or create token.")
 		return
 	}
```
You're not using the error here. Suggest to instead return a boolean like so:

```go
token, ok := getOrPostTokenWritesError(c, client, i)
if !ok {
	fmt.Println("Unable to get or create token.")
	return
}
```

This happens a few times throughout the file. Apply appropriately :)

applejag replied:
It tageted the old code instead of the new code in the preview, but the code block in my comment should be enough context I hope

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
 
-func getBodyFromRequest(user string, token string, url string) ([]byte, error) {
-	req, err := http.NewRequest(http.MethodGet, url, nil)
+func putProjectWritesError(c *gin.Context, client wharfapi.Client, provider wharfapi.Provider,
```
Suggest naming these functions `...WritesProblem` instead to make it more clear as we're only using the term "problem" when we refer to the `problem.Response`.

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
 	if err != nil {
-		fmt.Println("Unable to parse provider url: ", err)
-		return "", err
+		return importBody{}, err
```
This function seems excessive. Suggest inlining it as the caller needs to perform the same checks.

Also, now that we're using custom problem responses then [you should use `c.ShouldBindJSON` instead](https://iver-wharf.github.io/#/development/gin-shouldbind-vs-bind) :)

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
-		fmt.Println("Unable to get. Status code: ", resp.StatusCode)
-		return []byte{}, err
-	}
+	projectInDb, err := client.PutProject(
```
[Initialism](https://github.com/golang/go/wiki/CodeReviewComments#initialisms): `projectInDb` &rarr; `projectInDB`

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
+			BuildDefinition: buildDefinitionStr,
+			Description:     project.Description,
+			ProviderID:      provider.ProviderID,
+			GitURL:          gitURL})
```
nit: Suggest placing the `wharfapi.Project{` line on the same line as `client.PutProject` to reduce indentation.

```suggestion
	projectInDb, err := client.PutProject(wharfapi.Project{
		Name:            project.Name,
		TokenID:         i.TokenID,
		GroupName:       i.Group,
		BuildDefinition: buildDefinitionStr,
		Description:     project.Description,
		ProviderID:      provider.ProviderID,
		GitURL:          gitURL})
```

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
-func getBodyFromRequest(user string, token string, url string) ([]byte, error) {
-	req, err := http.NewRequest(http.MethodGet, url, nil)
+func putProjectWritesError(c *gin.Context, client wharfapi.Client, provider wharfapi.Provider,
+	i importBody, project azureDevOpsProject) (wharfapi.Project, error) {
```
I find this number of function arguments to be difficult to read.

You're passing along the `gin.Context`, `wharfapi.Client`, and `importBody` in most of these functions. Suggest to maybe create an ephemeral type where you have these functions as methods instead.

For example:

```go
type importer struct {
	ctx    *gin.Context
	client wharfapi.Client
	body   importBody
}

func (i importer) putProjectWritesError(provider wharfapi.Provider, project azureDevOpsProject) {}
func (i importer) postBranchesWritesError(projectInDB wharfapi.Project, project azureDevOpsProject) {}
```

And then maybe look at if you REALLY need the `importBody` in all the functions as you're passing it along almost everywhere, or if it could be elided from the proposed `importer` struct and only fed to those functions that really need it and not just the functions that pass it along because it's needed further down the line. It makes the functions less decoupled.

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
-		Value []azureDevOpsRef `json:"value"`
-		Count int              `json:"count"`
-	}{}
+// getTokenByIDWritesError Writes an error to the gin Context when an error occurs!
```
1. The function name is the first word in the sentence, so "Writes" should be lowercased here.

2. I advice against using exclamation marks for errors, even in comments. What's utterly important for you, which may be why you added the exclamation mark, may be totally irrelevant to some other coder while the exclamation mark drives focus towards it. Errors rarely need such focus. I find it better to reserve exclamation marks to cases such as "Remember to close the returned socket!" to not grow an exclamation-mark-fatigue

3. No need to document unexported functions. The function name should be self-explanatory, and here I don't think the comment adds much value to someone who already can easily access the source code. Also, as we don't comment unexported types regularly in Wharf, they get easily forgotten during refactorings and then get outdated quite quickly.

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
+	if err != nil || token.TokenID == 0 {
+		fmt.Printf("Unable to get token. %+v", err)
+		ginutil.WriteAPIClientReadError(c, err,
+			fmt.Sprintf("Unable to get token by id %v.", tokenID))
```
```suggestion
			fmt.Sprintf("Unable to get token by ID %d.", tokenID))
```

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
+		})
+		if err != nil {
+			fmt.Println("Unable to post branch: ", err)
+			ginutil.WriteAPIClientWriteError(c, err, fmt.Sprintf("Unable to import branch '%s'", branch.Name))
```
nit: There's a handy `%q` format that I personally enjoy which quotes strings for you. It will escape quotes in the string, but I think it's fine, especially as some of these problems might get reported back to us developers where escaped text can be useful.

```suggestion
			ginutil.WriteAPIClientWriteError(c, err, fmt.Sprintf("Unable to import branch %q", branch.Name))
```

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
+func getOrPostTokenWritesError(c *gin.Context, client wharfapi.Client, i importBody) (wharfapi.Token, error) {
+	var err error
+	if i.User == "" && i.TokenID == 0 {
+		err = fmt.Errorf("both token and user were omitted")
```
nit: Use `errors.New` when you're not doing any formatting

```suggestion
		err = errors.New("both token and user were omitted")
```

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
-			Name: name,
-			Ref:  ref.Name,
-		})
+	var token wharfapi.Token
```
nit: This `var token wharfapi.Token` and the above `var err error` seem excessive. Suggest to use declare the variables using `:=` where they're used instead.

I realize that most of this code is just moved, therefore the `nit:`

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
-	return projectBranches, nil
+	i.User = token.UserName
+	i.Token = token.Token
+	i.TokenID = token.TokenID
```
Possible regression: These assignments does nothing as `importBody` is a struct, and here you're only changing the local struct in this function and then not returning it.

I advice against just changing the parameter to a pointer as IMO it's not clear that the function takes a pointer based on its current name. Suggest to make sure it's returned properly instead, or use the `wharfapi.Token` that you're already returning as the state in the other functions instead of mutating the `importBody` type.

This is found in some of the other functions as well, so you might have to do a bigger refactor than you intentionally aimed for :/ You're free to do so, but it does enlarge the scope of this PR again which is a decision you'll have to make. You could chat with @fredx30 as I think he just went through a refactor for the wharf-provider-github repo in a similar way.

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
+		if err != nil || provider.ProviderID == 0 {
+			fmt.Printf("Unable to get provider. %+v", err)
+			ginutil.WriteAPIClientReadError(c, err,
+				fmt.Sprintf("Unable to get provider by id %v", i.ProviderID))
```
nit:

```suggestion
				fmt.Sprintf("Unable to get provider by ID %d.", i.ProviderID))
```

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
+}
+
+func getRepositoriesWritesError(c *gin.Context, i importBody, project azureDevOpsProject) (azureDevOpsRepositoryResponse, error) {
+	const apiVersion string = "5.0"
```
nit: I think you should inline these `apiVersion` consts. It doesn't give much value being extracted like this, we already can see the name of it in the map below.

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
 	if err != nil {
 		fmt.Println("Unable to get url: ", err)
-		return []byte{}, err
+		ginutil.WriteInvalidParamError(c, err, "URL", fmt.Sprintf("Unable to parse URL %q", i.URL))
```
nit: Use the JSON names for parameter errors

```suggestion
		ginutil.WriteInvalidParamError(c, err, "url", fmt.Sprintf("Unable to parse URL %q.", i.URL))
```

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
+	if repositories.Count != 1 {
+		fmt.Println("One repository is required.")
+		ginutil.WriteAPIClientReadError(c, err,
+			fmt.Sprintf("There were %v repositories, we need it to be 1.",
```
nit:
```suggestion
			fmt.Sprintf("There were %d repositories, we need it to be 1.",
```

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
+	if repositories.Value[0].Project.ID != project.ID {
+		fmt.Println("Repository is not connected with project.")
+		ginutil.WriteAPIClientReadError(c, err,
+			fmt.Sprintf("Repository id (%v) and project id (%v) mismatch",
```
nit:

```suggestion
			fmt.Sprintf("Repository ID (%d) and project ID (%d) mismatch.",
```

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
+	}, "%s/%s/%s/%s/%s", i.Group, i.Project, apiRepositories, projectName, itemsPath)
 	if err != nil {
 		fmt.Println("Unable to get url: ", err)
+		ginutil.WriteInvalidParamError(c, err, "URL", fmt.Sprintf("Unable to parse URL %q", i.URL))
```
nit: Use the JSON names for parameter problems

```suggestion
		ginutil.WriteInvalidParamError(c, err, "url", fmt.Sprintf("Unable to parse URL %q", i.URL))
```

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
+	if err != nil {
+		fmt.Println("Unable to get build definition: ", err)
+		ginutil.WriteFetchBuildDefinitionError(c, err,
+			fmt.Sprintf("Unable to fetch build definition for '%s'", i.Project))
```
nit:

```suggestion
			fmt.Sprintf("Unable to fetch build definition for project %q.", i.Project))
```

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
-func buildURL(urlStr string, group string, project string) (string, error) {
-	urlPath, err := url.Parse(urlStr)
+func getProjectsWritesError(c *gin.Context, i importBody) (azureDevOpsProjectResponse, error) {
+	var getProjectsURL *url.URL
```
Unnecessary `var` declaration

```suggestion
```

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
+		err = requests.GetAndParseJSON(&projects.Value[0], i.User, i.Token, getProjectsURL)
 	} else {
-		urlPath.Path = fmt.Sprintf("%v/%v", group, apiProjects)
+		err = requests.GetAndParseJSON(&projects, i.User, i.Token, getProjectsURL)
```
This implementation looks like two functions in one. Suggest to instead extract them into two separate, such as `getProject` and `getProjects` and only compare the `i.Project != ""` once instead.

applejag commented on azuredevops.go:
```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
+
+	err = requests.GetAndParseJSON(&projectRefs, i.User, i.Token, urlPath)
+	if err != nil {
+		ginutil.WriteProviderResponseError(c, err, "Could be caused by invalid JSON data structure." +
+			"\nMight be the result of an incompatible version of Azure DevOps.")
```
Suggest specifying what action caused this problem in the details. And suggest removing the newline as no other problem detail has those. It will be just as fine with just a space between them instead.

applejag commented on pkg/requests/requests.go:
```diff
@@ -0,0 +1,78 @@
+	"net/url"
+)
+
+// ConstructGetURL creates a GET request from a base URL, a map of query
+// parameters, and a formatted path.
+func ConstructGetURL(
+	rawURL string, queries map[string][]string, format string, values ...interface{}) (*url.URL, error) {
+
+	urlPath, err := url.Parse(rawURL)
+	if err != nil {
+		return nil, err
+	}
+
+	urlPath.Path = fmt.Sprintf(format, values...)
+	var q url.Values = queries
+	urlPath.RawQuery = q.Encode()
+
+	return urlPath, nil
+}
```
This package seems unnecessarily generic and in my opinion its usage bloats the code in `azuredevops.go`, especially with the `interface{}` map and formats. Would much rather see something like this:

```go
package requests

func NewGetRepositories(projectGroup, projectName string) (*url.URL, error) {}
func NewGetFile(projectGroup, projectName, filePath string) (*url.URL, error) {}
func NewGetProject(projectGroup, projectName) (*url.URL, error) {}
func NewGetProjects(projectGroup string) (*url.URL, error) {}
func NewGetGitRefs(projectGroup, projectName, refsFilter string) (*url.URL, error) {}
```

The above sample of functions would then be responsible for escaping all URL sections, adding the appropriate `apiVersion` query parameters, and so on.

applejag commented on pkg/requests/requests.go:
```diff
@@ -0,0 +1,78 @@
+	return urlPath, nil
+}
+
+// GetAndParseJSON invokes a HTTP request with basic auth.
+// On success the response body will be unmarshalled as JSON.
+func GetAndParseJSON(result interface{}, user, token string, urlPath *url.URL) error {
```
no need to export this function.

Alexamakans replied:
Used outside of the requests package still.

applejag replied:
Ooops, sorry

applejag commented on pkg/requests/requests.go:
```diff
@@ -0,0 +1,78 @@
+	return string(body), nil
+}
+
+func getBodyFromRequest(user string, token string, urlPath *url.URL) ([]byte, error) {
+	url := urlPath.String()
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return []byte{}, err
+	}
```
Think a good alternative would be to create a new type, for example named `azureClient`, that you feed in the credentials once instead of having to add them in every request function

Alexamakans replied:
In 8d3c659 I've extracted most things from azuredevops.go into `importer.AzureDevOpsImporter`, for lack of a better name.

What do you reckon is the main task of an "importer" vs. the main task of a "client"? I'm having some trouble thinking of a good distinction between the two.

applejag replied:
Client: basic functions to talk to the Azure DevOps APIs. Such as GetFile, ListProjects, et.al.

Importer: Integration between the Wharf client and Azure DevOps client. It uses the basic functions from the clients to coordinate the data transfer.

Just to separate the responsabilities and de-bloat the code

Alexamakans replied:
I've done what I imagine you meant in 1e1c3d4, but it feels pretty messy.
If you have any suggestions on how to make it cleaner, I'm all ears!

applejag reviewed (changes requested):
Awesome refactor! Lots of notes though :)

Alexamakans committed:
Changed ...WritesError -> ...WritesProblem and return bool instead of error

Alexamakans committed:
Replace BindJSON calls with ShouldBindJSON

Alexamakans committed:
Removed bindImportDetails function in favor of inline code

Alexamakans committed:
Separated getProjectsWritesProblem functionality into two functions

Alexamakans committed:
Renamed projectInDb -> projectInDB

Alexamakans committed:
Changed id -> ID and fmt.Errorf -> errors.New

Alexamakans committed:
Applied some suggestions from code review

Alexamakans committed:
Moved declaration of some variables

Alexamakans committed:
Extracted functionality into internal/importer

Alexamakans committed:
Split importer into AzureImporter and AzureClient and moved to internal/azure

Alexamakans committed:
Removed unnecessary json tags

applejag commented on internal/azure/client.go:
```diff
@@ -0,0 +1,201 @@
+	UserName  string
+	URL       string
+	UploadURL string
+}
```
As far as responsabilities goes, I don't think this client should even import the `wharfapi` package, and surely not include the token or provider structs inside it. Besides, they seem to be unused.

applejag commented on internal/azure/client.go:
```diff
@@ -0,0 +1,201 @@
+	// used in refresh only
+	UserName  string
+	URL       string
+	UploadURL string
```
If this type is only responsible for accessing Azure DevOps APIs, then it should not include code or comments that belong to the wharf domain. Refreshing a project is not something that the azureClient should have to deal with itself.

applejag commented on internal/azure/importer.go:
```diff
@@ -0,0 +1,302 @@
+const (
+	apiProviderName         = "azuredevops"
+	buildDefinitionFileName = ".wharf-ci.yml"
+	apiProjects             = "_apis/projects"
```
This `apiProjects` constant belongs to the `requests` package. Suggest moving it there

applejag commented on pkg/requests/requests.go:
```diff
@@ -0,0 +1,155 @@
+
+// GetAndParseJSON invokes a HTTP request with basic auth.
+// On success the response body will be unmarshalled as JSON.
+func GetAndParseJSON(result interface{}, user, token string, urlPath *url.URL) error {
+	body, err := getBodyFromRequest(user, token, urlPath)
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal(body, &result)
+
+	return err
+}
```
Suggest naming this `GetUnmarshalJSON` or something similar, replacing `Parse` with `Unmarshal` as it's the nomenclature in Go to show that you expect a pointer that will be populated via reflection

applejag commented on pkg/requests/requests.go:
```diff
@@ -0,0 +1,155 @@
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return []byte{}, fmt.Errorf("unable to get. status code: %d", resp.StatusCode)
+	}
```
Suggest checking for all 2xx codes instead of only 200.

Also, for the error it might be nice to use the `resp.Status` as that also includes the status code name that was provided by the server. Such as `"200 OK"` or `"400 Bad Request"`

```suggestion
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return []byte{}, fmt.Errorf("unable to get. status code: %s", resp.Status)
	}
```

applejag commented on pkg/requests/requests.go:
```diff
@@ -0,0 +1,155 @@
+		return nil, err
+	}
+
+	urlPath.Path = fmt.Sprintf("%s/%s/%s", groupName, projectName, apiRepositories)
```
I find these `apiRepositories` and similar consts more confusing than helping. Suggest inlining the values directly into the format instead.

applejag commented on internal/azure/importer.go:
```diff
@@ -0,0 +1,302 @@
+		ginutil.WriteInvalidBindError(c, err,
+			"One or more parameters failed to parse when reading the request body for import details.")
+		return nil, false
+	}
```
With the argument of SRP, then I have some comments:

- The new `internal/azure` package seems to deal with:
  - Utility methods to talk to the Azure DevOps API in a type-safe fashion, which should be its own package
  - Utility type to use the Azure DevOps API and the Wharf API and transfer data between them, which should be its own package
  - Handling and validating of incoming Gin HTTP requests, which should be its own package or live inside the `main` package

- The `internal/azure/importer.go` file seems to manage:
  - wharf-provider-azuredevops' incoming JSON structs and how they're serialized (with the `` `json:"tokenId"` `` attributes), which should live inside the `main` package next to the Gin method that accepts that JSON body.
  - Azure DevOps JSON response models, which should live in the same package (or even file) as the Azure DevOps API wrapper

Would rather see importer.go and client.go live in separate packages

Alexamakans committed:
Decoupled importer and client further

Alexamakans committed:
Readded WritesProblem to azureImporter exported methods

Alexamakans committed:
Moved NewGetXXX from pkg/requests/requests.go -> internal/azureapi/client.go and unexported them

Alexamakans committed:
Fixed trigger handler comment and so it returns 401 on failure to authenticate

Alexamakans committed:
Fixed build errors when swagging

fredx30 reviewed (commented):
Looked over this again just now and the code still looks nice to me but im having a hard time finding good answers to my usuall questions. Why did we open this PR? What problem does this solve? How do we benefit from solving it?
I can trace this back to vauge mentions of error output and error descriptions. As this PR alone now has 83 comments and its not the only PR that exsists for this goal i think its fair to demand an exlanation of why this is not simple accept&merge ~10 line string change. If this exsits please link it to the PR and in the changelog's change description.

fredx30 reviewed (approved):
As per earlier discussion addnig my green check here for neat code. Have not done any analysis on intent.

Alexamakans committed:
Renamed importBody -> importData

Alexamakans committed:
Changed reference to importData -> importBody in comment

Alexamakans committed:
Updated CHANGELOG.md

Alexamakans reviewed (commented):
typo: Title for 21ff73b is supposed to be `Renamed importData -> importBody`.
